### PR TITLE
Enhance rust client

### PR DESCRIPTION
**Pull request by elenaf9**
Enhance the rust-client:
- Merge upstream main branch and adapt to `punchr.proto` changes
- Handle errors individually and report back the proper `HolePunchOutcome` and error messages
- Change `agent_version` to `punchr/rust-client/..` to match the go version `punchr/go-client/..`
- Track the individual hole-punch attempts (up to 3)
- Switch from relay v2 to relay v1 since most client currently don't support v2 yet.
- Run a `ROUNDS` (currently 10) number of hole-punching requests, with 1s delay inbetween

From the first couple of tests it seems to work against the go-punchr server. That being said, the success rate is rather low. I still need to run more tests.

Edit: Open TODOs: 
- [x] There seems to be an invalid field in the `TrackHolePunchRequest` when the hole-punch succeeded. Need to investigate this more.
- [x] Spawn a new swarm on each round to prevent connection reverse-dial through port-reuse

---

While working on this client, I noticed a couple of things about the rust-DCUtR protocol:
- we currently don't support reverse-dial (which is a known issue)
- unless I am missing something, we don't measure the rtt to calculate the optimal point for starting the dial-attempt as dialer
- the behaviour currently does not report to the user if individual hole-punch attempts fail (which can be up to 3 until the whole hole-punching process is considered to be failed)
- on the dialer side, no `DirectConnectionUpgradeFailed` is reported if the hole-punching (that was initiated by the remote) failed

I am happy to tackle these issue, but I wanted to check first if some of these were conscious decisions.

### CODE REVIEW

**Comment by elenaf9** [comment]
One more thing I noticed: We intentionally skip the hole-punch attempt to a peer if their relay is listening on a quic-transport, because we don't support that transport yet. We do not send back a `TrackHolePunchRequest` in that case.
But when we then send a new `GetAddrInfoReques` to the server, the server keeps sending back the information about the same peer. @dennis-tra is there a way to solve this from the server side, or do we have to sent back a `TrackHolePunchRequest` to the server in this case as well?

**Comment by dennis-tra** [comment]
We have just discussed it in the call but want to also note it here :) For now, just send a `TrackHolePunchRequest` and give it an outcome of `CANCELLED` with an empty `hole_punch_attempts` array.

**Comment by mxinden** [comment]
> unless I am missing something, we don't measure the rtt to calculate the optimal point for starting the dial-attempt as dialer

RTT is measured here, though this code path won't be run in the Rust Punchr client, given that the remote initiates the DCUtR process and not the Punchr client.

https://github.com/libp2p/rust-libp2p/blob/7a1147877a52c7150190acdef35676cdac370299/protocols/dcutr/src/protocol/outbound.rs#L78

Does that make sense? I am happy to explain this in more depth.

**Comment by mxinden** [comment]
> the behaviour currently does not report to the user if individual hole-punch attempts fail (which can be up to 3 until the whole hole-punching process is considered to be failed)

We are currently only (internally) tracking this as the DCUtR initiator, thus on the remote side and not the Punchr client side. That said, we could track it (internally) on both ends and expose it in the DCUtR behaviour events.

**Comment by mxinden** [comment]
> on the dialer side, no `DirectConnectionUpgradeFailed` is reported if the hole-punching (that was initiated by the remote) failed

As of today, this is hard to infer on the dialer side. The initiator (listener side) opens up a new stream for each retry. Though there is no way for the dialer to gauge whether this was the final retry. I guess we could change the spec to require retries to happen on the same channel. Once that channel is closed and there is no direct connection, the dialer knows that the whole attempt failed. There might be other options I am unaware of.

File: rust-client/src/main.rs [hunk 1]
```diff
@@ -54,233 +55,378 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
-    {
-        Err(())
-    } else {
+    }
+
+    for _ in 0..ROUNDS {
+        let request = tonic::Request::new(grpc::RegisterRequest {
+            client_id: local_peer_id.to_bytes(),
+            agent_version: agent_version(),
+            // TODO: Automatically fill protocol list instead of hard coding it.
```
Comment (mxinden) [inline rust-client/src/main.rs]: As a hack, you could do something along the lines of:

https://github.com/libp2p/rust-libp2p/blob/7a1147877a52c7150190acdef35676cdac370299/swarm/src/lib.rs#L1326-L1333

File: rust-client/src/main.rs [hunk 2]
```diff
@@ -54,233 +55,378 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
-
-    let identify_config = IdentifyConfig::new("/TODO/0.0.1".to_string(), local_key.public())
+    let transport = relay_transport
+        .upgrade(upgrade::Version::V1)
+        .authenticate(noise::NoiseConfig::xx(noise_keys).into_authenticated())
+        // TODO: Consider supporting mplex.
+        .multiplex(libp2p::yamux::YamuxConfig::default())
+        .boxed();
+
+    let identify_config = IdentifyConfig::new("/identify/0.0.1".to_string(), local_key.public())
```
Comment (mxinden) [inline rust-client/src/main.rs]: Ah, this is confusing. The first parameter is the `protocol_version`, though what is meant is not the identify protocol version, but the protocol version of the whole p2p network:

https://github.com/libp2p/specs/tree/master/identify#protocolversion

Not sure what to use here. Likely just what every IPFS node uses, as we are connecting to the IPFS network.

**Review by mxinden — APPROVED** [review]
This is a lot cleaner. Thanks @elenaf9!

I just got two small comments. I would suggest we merge here as this is an improvement in every way and for the sake of small (future) pull requests. What do you think?

**Comment by dennis-tra** [comment]
> unless I am missing something, we don't measure the rtt to calculate the optimal point for starting the dial-attempt as dialer

Am I right that this problem came about when you wanted to fill in [this](https://github.com/dennis-tra/punchr/blob/08a000770cd38459a30b96279786363b10d04cd8/punchr.proto#L119) field?

If so, also Marten had doubts about the meaningfulness of the client-side RTT measurement in the Go implementation. So we may be able to just drop this field 🤷‍♂️

File: rust-client/src/main.rs (hunk 1 above)
Comment (elenaf9) [reply rust-client/src/main.rs]: Thanks, done in a0cdee1.

**Comment by elenaf9** [comment]
>> the behaviour currently does not report to the user if individual hole-punch attempts fail (which can be up to 3 until the whole  hole-punching process is considered to be failed)
>
> We are currently only (internally) tracking this as the DCUtR initiator, thus on the remote side and not the Punchr client side. That said, we could track it (internally) on both ends and expose it in the DCUtR behaviour events.

I think for now it's okay to keep it as it is and instead track it in the rust punchr client. Maybe we can circle back to this discussion once we also have the quic transport and know how it interferes with the DCUtR implementation.

**Comment by elenaf9** [comment]
>> unless I am missing something, we don't measure the rtt to calculate the optimal point for starting the dial-attempt as dialer
>
> Am I right that [this](https://github.com/dennis-tra/punchr/blob/08a000770cd38459a30b96279786363b10d04cd8/punchr.proto#L119) problem came about when you wanted to fill in this field?
>
> If so, also Marten had doubts about the meaningfulness of the client-side RTT measurement in the Go implementation. So we may be able to just drop this field man_shrugging

Yes, that was the reason. If the information is not useful we can just remove it, but since it is optional anyway I can also just leave it empty. I am fine with either of the two options.

**Comment by elenaf9** [comment]
@mxinden From my side this is ready to be merged. I removed the github workflow (running clippy on the rust client) again since it did not seem to work. I'll add it in a separate PR directly on the upstream repo after the rust-client is merged.

**Comment by mxinden** [comment]
Thanks @elenaf9!

> I'll add it in a separate PR directly on the upstream repo after the rust-client is merged.

Sounds good to me. I usually just copy https://github.com/actions-rs/example/blob/master/.github/workflows/quickstart.yml

**Pull request merged**